### PR TITLE
XWIKI-21520: Allow each pre-edit checker to be skipped

### PR DESCRIPTION
**Pull request by manuelleduc**
Jira: https://jira.xwiki.org/browse/XWIKI-21520
Forum proposal: https://jira.xwiki.org/browse/XWIKI-21520

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationCheckerResult.java [hunk 1]
```diff
@@ -39,16 +41,35 @@ public class EditConfirmationCheckerResult
 
     private final boolean isError;
 
+    private final Serializable cache;
+
     /**
      * Constructs a new object with the specified message and error status.
+     * This constructor initializes the {@code cache} with a {@code null} value,
+     * meaning that the result can't be forced.
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationCheckerResult.java:49]: It cannot be forced at all, or if it's forced then the user will be asked again?

File: xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationCheckerResult.java [hunk 2]
```diff
@@ -39,16 +41,35 @@ public class EditConfirmationCheckerResult
...
+    {
+        this(message, isError, null);
+    }
+
+    /**
+     * Constructs a new object with the specified message, an error status, and a cache result object.
+     *
+     * @param message the message associated with the result
+     * @param isError the error status of the result, error when {@code true}, warning otherwise
+     * @param cache the cache value associated with the result
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationCheckerResult.java:64]: I would specify also here that cache is nullable and what's the behaviour when null is used.

File: xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-default/src/main/java/org/xwiki/model/validation/edit/internal/DefaultEditConfirmationCheckersManager.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                check.ifPresentOrElse(editConfirmationCheckerResult -> {
+                    if (!result.isError()) {
+                        session.setAttribute(cachedKey, editConfirmationCheckerResult.getCache());
+                    }
+                }, () -> session.removeAttribute(cachedKey));
+                Optional<EditConfirmationCheckerResult> checkResult = check
+                    .flatMap(r -> {
+                        // We ignore the result if this is not an error, the cache value is not null, and the  
+                        // cache value stored in session is equal to the cache value returned by the analysis.
+                        if (!r.isError() && forcedValue != null && Objects.equals(r.getCache(), forcedValue)) {
```
Comment (surli) [inline xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-default/src/main/java/org/xwiki/model/validation/edit/internal/DefaultEditConfirmationCheckersManager.java:102]: So my understanding is that this `getCache` value is actually not something you want to have in cache, but something you want to use for *invalidating* the cache, is that correct? I'm not sure the name properly reflect that.

File: xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationCheckerResult.java (hunk 1 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationCheckerResult.java]: Updated, see https://github.com/xwiki/xwiki-platform/pull/2595/files#diff-bf2ed29667be98ddff3766113c301ce068688c7f6d10af4ef4f9e2c1f5996776R51

File: xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationCheckerResult.java (hunk 2 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationCheckerResult.java]: Updated, see https://github.com/xwiki/xwiki-platform/pull/2595/files#diff-bf2ed29667be98ddff3766113c301ce068688c7f6d10af4ef4f9e2c1f5996776R70

File: xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-default/src/main/java/org/xwiki/model/validation/edit/internal/DefaultEditConfirmationCheckersManager.java (hunk 3 above)
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-default/src/main/java/org/xwiki/model/validation/edit/internal/DefaultEditConfirmationCheckersManager.java]: `cache` renamed in `skipValue` wdyt?

File: xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationScriptService.java [hunk 4]
```diff
@@ -65,18 +57,15 @@ public class EditConfirmationScriptService implements ScriptService
...
-            return new EditConfirmationCheckerResults();
-        }
+        return this.editConfirmationCheckersManager.check();
+    }
+
+    /**
+     * Force the last {@link EditConfirmationChecker} components checks. Persist their last results, new checks are
+     * skipped as long as they match the persisted results.
+     */
+    public void force()
```
[the remark this answers is not in the document]
Comment (manuelleduc) [reply xwiki-platform-core/xwiki-platform-model/xwiki-platform-model-validation/xwiki-platform-model-validation-api/src/main/java/org/xwiki/model/validation/edit/EditConfirmationScriptService.java:70]: Fixed, thanks!

**Review by tmortagne — APPROVED** [review]

**Pull request merged**